### PR TITLE
Initialize Terraform with CloudFront CDN and S3 origin

### LINKED ISSUE #2: Provision S3 bucket and CloudFront distribution for the app
Use Terraform to provision an S3 bucket and CloudFront distribution to serve the app.

Ultimately, we want to hook the Mastarm build tool into these resources for deployments. For background, see the Mastarm docs: https://github.com/conveyal/mastarm#deploy.

### PR DESCRIPTION
**Pull request by jeancochrane**
## Overview

Perform initial Terraform setup, including a CloudFront CDN and S3 origin for the app. In addition, set up the foundation for the staging site at https://echolocator.azavea.com.

Closes #2.

## Demo

Visit https://echolocator.azavea.com to see a hello world message:

screen shot 2019-01-23 at 1 18 00 pm

## Testing instructions

* Create an `echo-locator` AWS profile using access credentials for the Professional Services account:

```
aws configure --profile echo-locator
```

* Run Terraform plan: 

```
docker-compose -f docker-compose.ci.yml run --rm terraform ./scripts/infra plan
```

* Confirm that the plan comes out clean
* Visit https://echolocator.azavea and confirm:
    * Page loads properly
    * Certificate valid
    * Content is the same as http://echo-locator-staging-site-us-east-1.s3.amazonaws.com/index.html

### CODE REVIEW

File: deployment/terraform/cdn.tf [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  restrictions {
+    geo_restriction {
+      restriction_type = "none"
+    }
+  }
+
+  viewer_certificate {
+    acm_certificate_arn      = "${module.cert.arn}"
+    minimum_protocol_version = "TLSv1"
```
Comment (hectcastro) [inline deployment/terraform/cdn.tf]: I think that we can update this to be `TLSv1.2_2018`.

File: deployment/terraform/cdn.tf [hunk 2]
```diff
@@ -0,0 +1,75 @@
+resource "aws_cloudfront_distribution" "cdn" {
+  origin {
+    domain_name = "${module.origin.site_bucket}.s3.amazonaws.com"
+    origin_id   = "originEastId"
+  }
+
+  enabled             = true
```
Comment (hectcastro) [inline deployment/terraform/cdn.tf]: Minor, but when we create CloudFront distributions via the UI, IPv6 is usually enabled. We can enable it here via `is_ipv6_enabled`.

File: deployment/terraform/cdn.tf [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      restriction_type = "none"
+    }
+  }
+
+  viewer_certificate {
+    acm_certificate_arn      = "${module.cert.arn}"
+    minimum_protocol_version = "TLSv1"
+    ssl_support_method       = "sni-only"
+  }
+}
```
Comment (hectcastro) [inline deployment/terraform/cdn.tf]: It looks like this resource has a `tags` attribute now too. That would be helpful to add in the event that we need to do accounting based on project tags (if we stay in a shared AWS account like we are now).

File: deployment/terraform/cdn.tf [hunk 4]
```diff
@@ -0,0 +1,75 @@
+resource "aws_cloudfront_distribution" "cdn" {
+  origin {
+    domain_name = "${module.origin.site_bucket}.s3.amazonaws.com"
+    origin_id   = "originEastId"
+  }
+
+  enabled             = true
+  http_version        = "http2"
+  comment             = "EchoLocator (${var.environment})"
```
Comment (hectcastro) [inline deployment/terraform/cdn.tf]: Potentially useless nit, but is it `EchoLocator`? I don't know, but wanted to check.

File: deployment/terraform/dns.tf [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  tags {
+    Project     = "${var.project}"
+    Environment = "${var.environment}"
+  }
+}
+
+resource "aws_route53_record" "site" {
+  zone_id = "${aws_route53_zone.external.id}"
+  name    = "${var.r53_public_hosted_zone_name}"
+  type    = "A"
```
Comment (hectcastro) [inline deployment/terraform/dns.tf]: If we enabled IPv6, we'd have to add an AAAA record here too.

File: deployment/terraform/storage.tf [hunk 6]
```diff
@@ -0,0 +1,15 @@
+module "origin" {
+  source = "github.com/azavea/terraform-aws-s3-origin?ref=0.3.0"
```
Comment (hectcastro) [inline deployment/terraform/storage.tf]: Nice module reuse.

File: deployment/terraform/cdn.tf [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  default_root_object = "index.html"
+  retain_on_delete    = true
+
+  price_class = "PriceClass_100"
```
[the remark this answers is not in the document]
Comment (hectcastro) [reply deployment/terraform/cdn.tf]: This makes sense to me.

File: deployment/terraform/cdn.tf (hunk 4 above)
Comment (jeancochrane) [reply deployment/terraform/cdn.tf]: Potentially unrelated, but I just realized that this should probably be `"${var.project} (${var.environment})"`.

**Comment by jeancochrane** [comment]
Alright, this should be ready for another pass.

File: deployment/terraform/cdn.tf (hunk 4 above)
Comment (ddohler) [reply deployment/terraform/cdn.tf]: We've been using `ECHOLocator` or `ECHO Locator`.

Comment (jeancochrane) [reply deployment/terraform/cdn.tf]: Ooh nice, thanks @ddohler!

Comment (hectcastro) [reply deployment/terraform/cdn.tf]: Yeah, comment was purely about the spacing and casing. Going with one of the ones Derek suggests sounds good.

Comment (jeancochrane) [reply deployment/terraform/cdn.tf]: Got it, should be updated now.

**Pull request merged**